### PR TITLE
GHA steps to collect and upload heap dumps to debug UT OOM errors

### PR DESCRIPTION
**Pull request by abhishekrb19**
Some [UTs](https://github.com/apache/druid/actions/runs/10796374752/job/29955096914?pr=17027) have recently [been](https://github.com/apache/druid/actions/runs/10781889856/job/29904194227) failing with OOMs: https://github.com/apache/druid/actions/runs/10796374752/job/29955096914?pr=17027

```
Warning:  Tests run: 15, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 4.868 s -- in org.apache.druid.msq.test.CalciteUnionQueryMSQTest
[INFO] Running org.apache.druid.msq.test.CalciteSelectQueryMSQTest
java.lang.OutOfMemoryError: Java heap space
Dumping heap to java_pid21084.hprof ...
Heap dump file created [3035323731 bytes in 8.410 secs]
Terminating due to java.lang.OutOfMemoryError: Java heap space
```

So this PR adds some GHA steps to tarball and upload the heap dump if they exist.

This PR has:

- [x] been self-reviewed.

### CODE REVIEW

File: .github/workflows/reusable-unit-tests.yml [hunk 1]
```diff
@@ -117,6 +117,31 @@ jobs:
           MAVEN_PROJECTS: ${{ inputs.maven_projects }}
         run: ./.github/scripts/unit_tests_script.sh
 
+      # Check if .hprof dumps exist on failure
+      - name: Check for .hprof files
+        if: ${{ failure() }}
+        id: check_for_heap_dump
+        run: |
+          if ls /home/runner/work/druid/druid/*.hprof 1> /dev/null 2>&1; then
```
Comment (imply-elliott) [inline .github/workflows/reusable-unit-tests.yml:125]:
```suggestion
          if ls ${GITHUB_WORKSPACE}/*.hprof 1> /dev/null 2>&1; then
```

File: .github/workflows/reusable-unit-tests.yml [hunk 2]
```diff
@@ -117,6 +117,31 @@ jobs:
...
+            echo "::set-output name=found_hprof::true"
+          else
+            echo "::set-output name=found_hprof::false"
+          fi
+
+      # Tarball the heap dumps on failure and if the heap dump exists
+      - name: Collect tarball hprof dumps on failure
+        if: ${{ failure() && steps.check_for_heap_dump.outputs.found_hprof == 'true' }}
+        run: |
+          tar cvzf ./hprof-dumps.tgz /home/runner/work/druid/druid/*.hprof
```
Comment (imply-elliott) [inline .github/workflows/reusable-unit-tests.yml:135]:
```suggestion
          tar cvzf ${RUNNER_TEMP}/hprof-dumps.tgz ${GITHUB_WORKSPACE}/*.hprof
```

File: .github/workflows/reusable-unit-tests.yml [hunk 3]
```diff
@@ -117,6 +117,31 @@ jobs:
...
+
+      # Upload the heap dump tarball on failure and if the heap dump exists
+      - name: Upload hprof dumps to GitHub
+        if: ${{ failure() && steps.check_for_heap_dump.outputs.found_hprof == 'true' }}
+        uses: actions/upload-artifact@master
+        with:
+          name: Hprof-${{ inputs.group }} hprof dumps (Compile=jdk${{ inputs.build_jdk }}, Run=jdk${{ inputs.runtime_jdk }})
+          path: hprof-dumps.tgz
```
Comment (imply-elliott) [inline .github/workflows/reusable-unit-tests.yml:143]:
```suggestion
          path: ${{ runner.temp }}/hprof-dumps.tgz
```

File: pom.xml [hunk 4]
```diff
@@ -1778,6 +1778,9 @@
                             -Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager
                             -Daws.region=us-east-1 <!-- required for s3-related unit tests -->
                             -Ddruid.test.stupidPool.poison=true
+                            <!-- the path needs to come from a maven option or so -->
+                            -XX:OnOutOfMemoryError='chmod 644 /home/runner/work/druid/druid/*.hprof'
```
Comment (imply-elliott) [inline pom.xml:1782]: Will environment variables be respected here?
```suggestion
                            -XX:OnOutOfMemoryError='chmod 644 ${GITHUB_WORKSPACE}/*.hprof'
```

Comment (imply-elliott) [reply pom.xml:1782]: Or maybe it can be path relative so this works everywhere, not just in Github Actions?

Comment (abhishekrb19) [reply pom.xml:1782]: Yeah, I'm not sure how best to do this. Ideally it'd be some relative path like you're referring to

Comment (abhishekrb19) [reply pom.xml]: Updated to use `${project.parent.basedir}/target`

Comment (kgyrtkirk) [reply pom.xml]: please use `${maven.multiModuleProjectDirectory}` ; I think that property will not necessarily get to the project root

**Review by kgyrtkirk — APPROVED** [review]
project.parent.basedir is already used ; and the module hierarchy is usually 1 level...we could accept the same risks...
+1

**Pull request merged**